pkg/volumeclaimbinder/types.go

```diff
@@ -139,3 +146,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
 	bSize := bQty.Value()
 	return aSize <= bSize
 }
+
+// allPossibleMatchingAccessModes returns an array of AccessMode arrays that can satisfy a user's requested modes.
+// There are, essentially, 3 ways to request access modes:
+//
+//  RWO -- the most basic form of access (all storage volumes)
+//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
+//  RWO+ROX+RWX -- many writers (NFS)
+//
+// A user's request can be satisfied with more modes, but never fewer.
+// This func returns modes with ascending levels of modes to give the user what is closest to what they actually asked for.
+//
+func (pvIndex *persistentVolumeOrderedIndex) allPossibleMatchingAccessModes(requestedModes []api.PersistentVolumeAccessMode) [][]api.PersistentVolumeAccessMode {
+	matchedModes := [][]api.PersistentVolumeAccessMode{}
+	keys := pvIndex.Indexer.ListIndexFuncValues("accessmodes")
```

Comment:
Ok

On Fri, Jul 24, 2015 at 11:29 AM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/volumeclaimbinder/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10833#discussion_r35433497>
> :
>
> > @@ -139,3 +146,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
> >  	bSize := bQty.Value()
> >  	return aSize <= bSize
> >  }
> > +
> > +// allPossibleMatchingAccessModes returns an array of AccessMode arrays that can satisfy a user's requested modes.
> > +// There are, essentially, 3 ways to request access modes:
> > +//
> > +//  RWO -- the most basic form of access (all storage volumes)
> > +//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
> > +//  RWO+ROX+RWX -- many writers (NFS)
> > +//
> > +// A user's request can be satisfied with more modes, but never fewer.
> > +// This func returns modes with ascending levels of modes to give the user what is closest to what they actually asked for.
> > +//
> > +func (pvIndex *persistentVolumeOrderedIndex) allPossibleMatchingAccessModes(requestedModes []api.PersistentVolumeAccessMode) [][]api.PersistentVolumeAccessMode {
> > +	matchedModes := [][]api.PersistentVolumeAccessMode{}
> > +	keys := pvIndex.Indexer.ListIndexFuncValues("accessmodes")
>
> I wanted to account for all PVs that were actually added instead of
> attempting to derive all possible combinations myself. Asking for what it
> has indexed was the easiest way to find all current access modes in use.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10833/files#r35433497>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift